truffle/src/main/ruby/core/rubinius/api/shims/thread.rb

```diff
@@ -15,4 +42,43 @@ def recursive_objects
     @recursive_objects ||= {}
   end
 
+  def raise_prim(exc)
+    Rubinius.primitive :thread_raise
+    Kernel.raise PrimitiveFailure, "Thread#raise primitive failed"
+  end
+
+  private :raise_prim
+
+  def raise(exc=undefined, msg=nil, trace=nil)
+    return self unless alive?
+
+    if undefined.equal? exc
+      no_argument = true
+      exc         = nil
+    end
+
+    if exc.respond_to? :exception
+      exc = exc.exception msg
+      Kernel.raise TypeError, 'exception class/object expected' unless Exception === exc
+      exc.set_backtrace trace if trace
+    elsif no_argument
+      exc = RuntimeError.exception nil
+    elsif exc.kind_of? String
+      exc = RuntimeError.exception exc
+    else
+      Kernel.raise TypeError, 'exception class/object expected'
+    end
+
+    if $DEBUG
+      STDERR.puts "Exception: #{exc.message} (#{exc.class})"
+    end
+
+    if self == Thread.current
+      Kernel.raise exc
+    else
+      exc.capture_backtrace! 2 unless exc.backtrace?
+      raise_prim exc
+    end
+  end
+
```

Comment:
All of this should be in a separate file, because the loading order will load Rubinius' common/thread.rb after this according to [core.rb](https://github.com/jruby/jruby/blob/master/truffle/src/main/ruby/core.rb#L29).
The solution is to move it to a separate file, for instance named `thread_common.rb` or `thread_after_common.rb` and add that one after `common/thread,rb` in core.rb.
We should have a better organization for all of this.
cc @nirvdrum.